File: apps/satellite-service/src/middleware/request-id.ts
/**
 * @file middleware/request-id.ts
 * @description Injects a unique request ID into each request for distributed tracing.
 * Checks for an incoming X-Request-Id header (set by load balancer/ingress) first,
 * then generates a UUID v4 if absent. The ID is echoed back in the response header.
 */

import type { Request, Response, NextFunction } from 'express';
import { v4 as uuidv4 } from 'uuid';

declare global {
  // Augment Express Request type to include the requestId
  // eslint-disable-next-line @typescript-eslint/no-namespace
  namespace Express {
    interface Request {
      id: string;
    }
  }
}

/**
 * Express middleware that ensures every request has a unique ID.
 */
export function requestIdMiddleware(req: Request, res: Response, next: NextFunction): void {
  const incomingId = req.headers['x-request-id'];
  const requestId = typeof incomingId === 'string' ? incomingId : uuidv4();

  req.id = requestId;
  res.setHeader('X-Request-Id', requestId);

  next();
}

File: apps/planetary-service/src/middleware/request_id.py
"""Request ID propagation middleware."""

import uuid

from starlette.middleware.base import BaseHTTPMiddleware
from starlette.requests import Request
from starlette.responses import Response


class RequestIdMiddleware(BaseHTTPMiddleware):
    """Attach a unique X-Request-Id to every request/response."""

    async def dispatch(self, request: Request, call_next: object) -> Response:
        request_id = request.headers.get("X-Request-Id") or str(uuid.uuid4())
        response: Response = await call_next(request)  # type: ignore[arg-type]
        response.headers["X-Request-Id"] = request_id
        return response
